scripts/ci/crates.py

```diff
@@ -77,30 +77,39 @@ def crate_deps(member: Dict[str, Dict[str, Any]]) -> Generator[Tuple[str, Depend
             yield (v, DependencyKind.BUILD)
 
 
-def get_sorted_publishable_crates(crates: Dict[str, Crate]) -> Dict[str, Crate]:
+def get_sorted_publishable_crates(ctx: Context, crates: Dict[str, Crate]) -> Dict[str, Crate]:
     """
     Returns crates topologically sorted in publishing order.
 
     This also filters any crates which have `publish` set to `false`.
     """
 
     def helper(
+        ctx: Context,
         crates: Dict[str, Crate],
         name: str,
         output: Dict[str, Crate],
+        visited: Dict[str, bool],
     ) -> None:
         crate = crates[name]
         for dependency, _ in crate_deps(crate.manifest):
             if dependency not in crates:
                 continue
-            helper(crates, dependency, output)
+            helper(ctx, crates, dependency, output, visited)
         # Insert only after all dependencies have been traversed
-        if "publish" not in crate.manifest["package"] or crate.manifest["package"]["publish"]:
-            output[name] = crate
-
+        if name not in visited:
+            visited[name] = True
+            if "publish" not in crate.manifest["package"]:
+                ctx.error(
+                    f"Crate {Fore.BLUE}{name}{Fore.RESET} does not have {Fore.BLUE}package.publish{Fore.RESET} set."
+                )
+            elif crate.manifest["package"]["publish"]:
+                output[name] = crate
```

Comment:
tip: don't put negations in `if`-statements if there is an `else` statement. Flip it around and avoid unnecessary negations!

Reply:
i used `.get("publish")` here instead to make it a bit clearer